### ISSUE TITLE
[GSoC] [draft] Implement widgets for toga_winforms

### ISSUE THREAD

**Issue by cxue4ur19** [pybee/toga#419]
**Overview**
Compared to core/toga/widgets (the frontend API), toga_winform/widgets (backend API for windows) is missing the following:

- Canvas
- Detailed list
- Image
- Imageview
- Selection
- Slider
- Switch
- Tree

Also, font.py is not implemented for toga_winform

During GSoC, I hope to improve WinForms backend by implementing above widgets for toga_winform and improving dialog support.

Obulat is contributing to winforms at a high rate, but I think I can help to accelerate this process. I have no rigid commitments for the summer, so I will be able to contribute 40 hr/week. 

**Timetable**  
## → Getting familiar with things - 1 week 
1.1 Read toga code in more detail, in order to:
make sure I understand what files are related to the work I’ll be doing, as well as why and how they are related  
Get used to BeeWare code style conventions
1.2  Ask questions in Gitter for clarifications 
1.3  Read winForm documentation and Windows Human Interface guide more in depth
1.4  Run windows cookiecutter template to get ready for widget testing after implementation of each widget

→ Implementing missing widgets - 9 weeks [not necessarily following the order]

**Quick evaluation of a existing widget for reference: Button**
WHAT: specified by documentation and functions in core/toga
A clickable button that has a string label, a id for identification, a function that would be invoked when pressed, a optional style, a boolean defining whether it can be interacted with
WINFORM CLASS AND DEFAULT PROPERTIES OR METHODS USED
Class: System.Windows.Forms.Button
Methods and Properties: 
The following defined whether button can be interacted with 
[P]Enabled: get/set a value indicating control-ability 
The following defined the style:
[P]BackColor: get/set background color of the control (means widget)
[P]Text: get/set text associated with the control
[P]PreferredSize: get/set size associated with the control, has property height and width in pixels, both were used 
No need to define a function to handle what happens when button is pressed 

## 2.1 Detailed List - 1.5 week
2.1.1 WHAT: specified by documentation and functions in core/toga
Optionally styled widget that holds a list of strings in a vertical list format, the list can be refreshed. Each string occupies a row, these rows can be deleted and selected. Two scroll methods defined that would scroll the view until row with given index is visible. 
2.1.2 POTENTIAL WINFORMS CLASS AND DEFAULT METHODS 
Class:System.Windows.Forms.ListBox [displays a list of items]
Methods and Properties:
[P]AllowSelection: get a value indicating whether currently enables selection of list items
[P]SelectionMode: can enable multiple selection by setting this to SelectionMode.MultiExtended
[P]BackColor: get/set
[P]ColumnWidth: get/set
[P]Items (belongs to ListBox.ObjectCollection): a list of objects, can be strings, has methods Add() and RemoveAt(); both methods takes a single argument index, which can be defined by [P]SelectedIndex or [P] SelectedIndices depending on selection mode
[M]Refresh(): used to refresh items
2.1.3 Risks: 
- I think being able to do multi-select would be cool, but the documentation for the widget is missing, so I do not know if other platforms implement this function; I can overcome this by reading toga_cocoa
- No default for scrolling until some row is visible
- I did not find a default winForms method that could match scroll_to_row defined in core/toga, more research is needed to figure how how to implement these
- I need to think about how to define the way refresh can be triggered, is there a toga convention [i.e. when dragged down from top/when a button pressed], or is this left to be defined by app developers; can overcome by reading toga_cocoa

[note: other widgets will be updated soon]
2.2 Image and ImageView - 2 weeks
Core: Image contains a path to an image, ImageView displays the image with that path
WinForms: use PictureBox [allows bitmap,GIF, JPEG, metafile, icon]
Risk: from personal experience, working with image is tricky, there is always a new way for something to go wrong; so allow extra time for this task

2.3 Canvas implementation - 1.5 week
Core: there are about 21 methods that need to be defined by backend
WinForms: use Canvas, there are default methods that map to each method above

2.4 Tree implementation - 1.5 week
Core: holds data in a tree structure
WinForms: use TreeView [allows user to view and interact with hierarchically arranged collection of objects]
Risk: according to toga/docs, this is still under development, so the task might be more difficult than I am expecting

2.5 Selection implementation - 0.5 week
Core: pick one option from some defined string selections
WinForms: use Command Links [select a single option to a instruction and an action is immediately carried out]

2.6 Slider implementation - 0.5 week
Core: display a range of values
WinForms: use TrackBar

2.7 Switch Implementation -0.5 week
Core: a button that is in on state when checked and in off state when unchecked
WinForms: use CheckBox and set appearance to Button to disguise it as a switch

## →  Improving “Messages” - 2 weeks
Windows’ guideline defines four types of message users need/want to see when using an application: errors, warning, confirmations, and notifications.  Currently, toga_winforms/dialogs.py has some basic implementations but I think it can do better. 
_Current State_
There are three methods (question, confirm, error) that invoke MessageBox class and show:
- question: message, title, Yes/No buttons
- confirm: message, title, Ok/Cancel buttons
- error: message, titel, Ok button, error icon
_To Do_
[1 week]
- toga_coacoa’s implementation of question and confirm has style and icon, make toga_winforms’ implementation consistent with it
- Add implementation for warning
- Add implementation for notification
- Investigate whether other platforms can also implement warning and notification, if so, how
[1 week]
- There are two methods that invokes window.platform.not_implemented (save_file, stack_trace):
- investigate why they are not implemented
- if they can be implemented, them make implementations consistent with toga_cocoa
- Risk: core/toga does not expose dialog, why? 

_Testing_
After implementation of each widget:
Human testing: use the widget in the demo app prepared on week one; the widget should look natural for users and easy to use for developers
Unit testing: I need to look deeper into toga_dummy/test_implementation.py to understand exactly what exactly my widgets would do with unit testing. 

**Risks**
1. Limited experience using Toga as a user
- It might be a good idea to see what GUI in iOS looks like as an user in order to have a reference of what correctly implemented widgets should look like
2. I do not know how different Windows phone and Windows desktop are in terms of GUI, this might be something I need to look into; however, from what I have seen, Windows phones are not popular, so I am not sure if it is worth the time
3. Likely need to take the GRE sometime in August

**About Me**
My name is Chaoying Xue. I am currently a undergraduate student studying Computer Science and English Literature in Rochester, New York. My time zone is EST. 

References:
[.Net API Browser](https://docs.microsoft.com/en-us/dotnet/api/)
[Windows Guidelines](https://msdn.microsoft.com/en-us/library/windows/desktop/dn688964(v=vs.85).aspx)

**Comment by freakboy3742** [comment]
This is a good start, and it was on a really good track until the last couple of weeks of the timeline :-)

The "2 week" block for image view is probably reasonable, but the fact that you're stating a large period, and openly admit that it will be "tricky" is an indicator that more research is required. 

The "3 week" block right at the end reads as a "I'll do some stuff" block. Google (and us) are looking for a clear plan that you're going to make use of the full 12 weeks. It doesn't matter so much if you don't actually end up following the plan exactly, but we are looking for an initial intent to use the full 12 weeks.

Lastly, some of the widgets you're referenced are really simple (like Switch), and some (like detailed list and canvas) involve some fairly complex data management and/or graphics API management. You've definitely allocated *more* time for the complex widgets - but this proposal would be much stronger if it included details on the specific aspects that will need to be tackled, and gave an indication of how you would manage the complexity.

**Comment by cxue4ur19** [comment]
Thank you for the review, it was really helpful!

I improved my proposal regarding detailedlist and the "3 week block"; I also added a quick review of a already an already implemented widget (Button) as reference for future work. There will be another update regarding other widgets, especially image and imageView.  

I decided to add a partial update because some of the claims I made in the original draft do not apply anymore after I did more reading. Of course, I also hope to get some feedback on whether the kind of changes I am making is helping the proposal since it is almost the weekend.

**Comment by freakboy3742** [comment]
This is definitely an improvement. The formatting is a bit hard to follow (I'm guessing that might be because it's been copied from a word processor of some kind).

I still have some concerns about your timings; mostly because there are still blocks up of up to 2 weeks, and we don't have any basis for evaluating how realistic your estimates are. However, the individual estimates are internally consistent (e.g., simple widgets are all 0.5 weeks), so I guess that means the schedule will either be massively over or under budget, depending on how much progress you actually end up making.

There probably isn't enough time for another review at this point, so cleaning up the formatting and  submitting a final version is probably the best course of action at this point.

**Comment by freakboy3742** [comment]
Unfortunately, this project was not selected for the 2018 GSoC. Thanks for applying!
